src/platform/silabs/SiWx917/OTAImageProcessorImpl.cpp

```diff
@@ -258,20 +264,49 @@ void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
         if (writeBufOffset == kAlignmentBytes)
         {
             writeBufOffset = 0;
-
-            CORE_CRITICAL_SECTION(err = bootloader_eraseWriteStorage(mSlotId, mWriteOffset, writeBuffer, kAlignmentBytes);)
-            if (err)
+            if(flag == RPS_HEADER)
             {
-                ChipLogError(SoftwareUpdate, "ERROR: In HandleProcessBlock bootloader_eraseWriteStorage() error %ld", err);
-                imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
-                return;
+                // Send RPS header which is received as first chunk
+                status = sl_si91x_fwup_start(writeBuffer);
+                status = sl_si91x_fwup_load(writeBuffer, kAlignmentBytes);
+                total_size = imageProcessor->mParams.totalFileBytes;
+                count = count + kAlignmentBytes;
+                flag = RPS_DATA;
             }
-            mWriteOffset += kAlignmentBytes;
-            imageProcessor->mParams.downloadedBytes += kAlignmentBytes;
-        }
-    }
-
-    imageProcessor->mDownloader->FetchNextData();
+            else if(flag == RPS_DATA)
+            {
+                // Send RPS content
+                count = count + kAlignmentBytes;
+                if(count >= total_size)
+                {
+                    length = kAlignmentBytes - (count - total_size);
+                    last = 1;
+                }
+                else
+                {
+	                length = kAlignmentBytes;
+                }
+                status = sl_si91x_fwup_load(writeBuffer, length);
+	            if (status != 0) {
+					ChipLogProgress(SoftwareUpdate, "status: 0x%lX", status);
+                    if (status == SL_STATUS_FW_UPDATE_DONE) {
+                        ChipLogProgress(SoftwareUpdate, "M4 Firmware update complete");
+                        // send system reset request to reset the MCU and upgrade the m4 image
+                        osDelay(5000);
+                        sl_si91x_soc_soft_reset();
```

Comment:
What is the difference between this fw update complete/soft reset and the one in `HandleFinalize` line 187 -190 ?

Reply:
I have modified the file according to your comments.